proposals/0000-th-stage-hygiene.rst

```diff
@@ -0,0 +1,397 @@
+Stage Hygiene for Template Haskell
+==============
+
+.. proposal-number:: Leave blank. This will be filled in when the proposal is
+                     accepted.
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/243>`_.
+.. sectnum::
+.. contents::
+
+Template Haskell currently doesn't work well with cross compilation.
+This is a big nuisance for anyone making phone or browser software.
+To fix this, enforce stronger separation between the stages.
+
+Motivation
+------------
+
+In the olden days, one couldn't use Template Haskell with cross compilation at all.
+This is bad in itself, but also bad from an ecosystem perspective.
+Haskell has an excellent culture of code reuse, but that means that arbitrary software is very likely to depend on Template Haskell.
+That lack of being able to write idiomatic Haskell code with popular libraries for all platform is as big of a loss as not being able to Template Haskell itself.
+The risk of library fragmentation between between those writing code for weird platforms and those not is also bad for everyone.
+
+The external interpreter made it at least possible, relatively automatically, but doesn't work if you explicitly want side effects to be run on the compiler's platform.
```

Comment:
You only mention TH's side-effects here, afaict. That can matter a lot. I assume the idea is that side-effects will run where they are compiled, so that code which is compiled for the host will run on the host etc?

An extension which I have considered for a while (which would be possible following https://github.com/ghc-proposals/ghc-proposals/pull/246) would be to have something like `PureTemplateHaskell`, which would restrict the type of splices to be `NameSupply m => ...` rather than `Quasi m => ...` i.e. ban IO from quotes. This would play quite nicely with this proposal, I think.

Reply:
Yes that dovetails. I think it's good to separate the stages in any event, it's also nice to turn off effects. With either enabled, the other is just a "human convenience", but a good one.

Reply:
Sure, but I do think you should explicitly say what happens with TH side effects in this proposal.